File: link_files.php
<?php
/**
 * Deduplikace + propojení souborů s písněmi
 * Spustit: php link_files.php
 */
require_once "config.php";

$raw = file_get_contents($LOCAL_DB);
$data = json_decode($raw, true);
if (!is_array($data)) die("Chyba čtení songs.json\n");

// Záloha
file_put_contents($LOCAL_DB . ".bak", $raw);
echo "Záloha vytvořena.\n\n";

// Normalizace názvu pro porovnání
function norm($s) {
    $s = mb_strtolower(trim($s), 'UTF-8');
    // Odstraň příponu souboru
    $s = preg_replace('/\.(pdf|xml|sng|txt)$/i', '', $s);
    // Odstraň přídavky typu "- pěvecký sbor", "_kompakt", "_méně akordů", "_rozepsano", " 2"
    $s = preg_replace('/[\s_-]+(pěvecký sbor|kompakt|méně akordů|rozepsano|nevánoční verze)$/u', '', $s);
    $s = preg_replace('/\s+\d+$/', '', $s); // trailing číslo ("Zachránce vzácný 2")
    // Odstraň interpunkci (čárky, tečky, pomlčky...)
    $s = str_replace(['_', '-', ',', '.', '!', '?', ':', ';', '–', '—', '\'', '"'], ' ', $s);
    // Odstraň nadbytečné mezery
    $s = preg_replace('/\s+/', ' ', trim($s));
    return $s;
}

// === 1. DEDUPLIKACE ===
echo "=== DEDUPLIKACE ===\n";
$unique = [];
$merged = 0;

foreach ($data as $song) {
    $key = norm($song['name']);
    
    if (isset($unique[$key])) {
        $e = &$unique[$key];
        // Sloučit historii
        $h = array_unique(array_merge($e['history'] ?? [], $song['history'] ?? []));
        rsort($h);
        $e['history'] = array_values($h);
        $e['count'] = count($e['history']);
        if (count($e['history']) > 0) $e['last'] = $e['history'][0];
        // Doplnit co chybí
        foreach (['author','category','tempo','tags','pdf','openlp'] as $f) {
            if (empty($e[$f]) && !empty($song[$f])) $e[$f] = $song[$f];
        }
        echo "  Sloučeno: '{$song['name']}' → '{$e['name']}'\n";
        $merged++;
        unset($e);
    } else {
        $unique[$key] = $song;
    }
}
$data = array_values($unique);
echo "Sloučeno: $merged duplicit\nPísní po deduplikaci: " . count($data) . "\n\n";

// === 2. PROPOJENÍ SOUBORŮ ===
echo "=== PROPOJENÍ SOUBORŮ ===\n";

// Buildneme lookup: norm(songname) → index
$lookup = [];
foreach ($data as $i => $song) {
    $lookup[norm($song['name'])] = $i;
}

function linkFiles($dir, $field, &$data, &$lookup) {
    if (!is_dir($dir)) return;
    $linked = 0;
    $unmatched = [];
    
    foreach (scandir($dir) as $f) {
        if ($f === '.' || $f === '..') continue;
        $key = norm($f);
        
        if (isset($lookup[$key])) {
            $idx = $lookup[$key];
            if (empty($data[$idx][$field])) {
                $data[$idx][$field] = $f;
                $linked++;
                echo "  [{$field}] '{$data[$idx]['name']}' ← $f\n";
            }
        } else {
            $unmatched[] = $f;
        }
    }
    
    echo "  Propojeno: $linked\n";
    if ($unmatched) {
        echo "  Nepropojené:\n";
        foreach ($unmatched as $u) echo "    - $u\n";
    }
}

echo "\n--- PDF ---\n";
linkFiles(__DIR__ . "/pdf/", "pdf", $data, $lookup);

echo "\n--- OpenLP ---\n";
linkFiles(__DIR__ . "/openlp/", "openlp", $data, $lookup);

// Uložit
file_put_contents($LOCAL_DB, json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
echo "\n✅ Hotovo! songs.json aktualizován.\n";

File: link_files.py
#!/usr/bin/env python3
"""
Deduplikace + propojení souborů s písněmi v songs.json
Spustit: python3 link_files.py
"""
import json, os, re, shutil

BASE = os.path.dirname(os.path.abspath(__file__))
DB = os.path.join(BASE, "data", "songs.json")
PDF_DIR = os.path.join(BASE, "pdf")
OPENLP_DIR = os.path.join(BASE, "openlp")

with open(DB, "r", encoding="utf-8") as f:
    data = json.load(f)

# Záloha
shutil.copy2(DB, DB + ".bak")
print("Záloha vytvořena.\n")

def norm(s):
    """Normalizace pro porovnání (case-insensitive, bez interpunkce, bez přídavků)"""
    s = s.lower().strip()
    # Odstraň příponu
    s = re.sub(r'\.(pdf|xml|sng|txt)$', '', s, flags=re.I)
    # Odstraň přídavky
    s = re.sub(r'[\s_-]+(pěvecký sbor|kompakt|méně akordů|rozepsano|nevánoční verze)$', '', s)
    s = re.sub(r'\s+\d+$', '', s)  # trailing číslo
    # Odstraň interpunkci
    s = re.sub(r'[_\-,\.!?:;–—\'\"()]', ' ', s)
    s = re.sub(r'\s+', ' ', s).strip()
    return s

# === 1. DEDUPLIKACE ===
print("=== DEDUPLIKACE ===")
unique = {}
merged = 0

for song in data:
    key = norm(song["name"])
    if key in unique:
        e = unique[key]
        # Sloučit historii
        h = list(set((e.get("history") or []) + (song.get("history") or [])))
        h.sort(reverse=True)
        e["history"] = h
        e["count"] = len(h)
        if h:
            e["last"] = h[0]
        # Doplnit prázdná pole
        for f in ["author", "category", "tempo", "tags", "pdf", "openlp"]:
            if not e.get(f) and song.get(f):
                e[f] = song[f]
        print(f"  Sloučeno: '{song['name']}' → '{e['name']}'")
        merged += 1
    else:
        unique[key] = song.copy()

data = list(unique.values())
print(f"Sloučeno: {merged} duplicit")
print(f"Písní po deduplikaci: {len(data)}\n")

# === 2. PROPOJENÍ SOUBORŮ ===
print("=== PROPOJENÍ SOUBORŮ ===")

# Lookup: norm(název) → index
lookup = {norm(s["name"]): i for i, s in enumerate(data)}

def link_files(directory, field):
    if not os.path.isdir(directory):
        print(f"  Složka {directory} neexistuje!")
        return
    linked = 0
    unmatched = []
    for f in sorted(os.listdir(directory)):
        if f.startswith('.'):
            continue
        key = norm(f)
        if key in lookup:
            idx = lookup[key]
            if not data[idx].get(field):
                data[idx][field] = f
                linked += 1
                print(f"  [{field}] '{data[idx]['name']}' ← {f}")
        else:
            unmatched.append(f)
    
    print(f"  Propojeno: {linked}")
    if unmatched:
        print("  Nepropojené:")
        for u in unmatched:
            print(f"    - {u}  (norm: '{norm(u)}')")

print("\n--- PDF ---")
link_files(PDF_DIR, "pdf")

print("\n--- OpenLP ---")
link_files(OPENLP_DIR, "openlp")

# Uložit
with open(DB, "w", encoding="utf-8") as f:
    json.dump(data, f, ensure_ascii=False, indent=4)
print(f"\n✅ Hotovo! songs.json aktualizován ({len(data)} písní).")
